### PR TITLE
[tool] Add support for `.java`, `.gradle`, `.sh`, and `.m` files…

### PR DESCRIPTION
**Pull request by mike-v2**
… to `update-excerpts` tool

Need this tool to recognize `.java`, `.gradle`, and `.sh` files in order to complete [flutter/flutter#102679](https://github.com/flutter/flutter/issues/102679). The `README.md` for the `espresso` package contains code from `.java`, `.gradle`, and `.sh` files.

Support for `.m` files was requested by @tarrinneal.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

**Comment by mike-v2** [comment]
Ah sorry, I should've figured that out. Thanks for pointing me in the right direction!

File: script/tool/test/update_excerpts_command_test.dart [hunk 1]
```diff
@@ -447,6 +443,74 @@ FAIL
...
+
+<?code-excerpt "$fileName (SomeSection)"?>
+```$language
+A B C
+```
+''', fileName);
+      });
+    }
+
+    // ignore: prefer_foreach
```
Comment (mike-v2) [inline script/tool/test/update_excerpts_command_test.dart:509]: Two conflicting lint rules here: [avoid_function_literals_in_foreach_calls](https://dart.dev/tools/linter-rules/avoid_function_literals_in_foreach_calls) and [prefer_foreach](https://dart.dev/tools/linter-rules/prefer_foreach). The latter aims to make the code more terse but the formatter uses the same number of lines with `forEach` as it does with `for...in`.

File: script/tool/lib/src/update_excerpts_command.dart [hunk 2]
```diff
@@ -142,10 +142,15 @@ class UpdateExcerptsCommand extends PackageLoopingCommand {
                   case '.kt':
                     language = 'kotlin';
                     break;
-                  case '.cc':
```
Comment (stuartmorgan) [inline script/tool/lib/src/update_excerpts_command.dart:145]: Why was this removed?

File: script/tool/test/update_excerpts_command_test.dart [hunk 3]
```diff
@@ -33,12 +33,13 @@ void runAllTests(MockPlatform platform) {
       ));
   });
 
-  Future<void> testInjection(String before, String source, String after,
+  Future<void> testInjection(
+      String before, String source, String after, String fileName,
```
Comment (stuartmorgan) [inline script/tool/test/update_excerpts_command_test.dart:37]: This is a lot of positional arguments with the same type; ideally they should all be named, but lets at least name the new one.

Comment (stuartmorgan) [inline script/tool/test/update_excerpts_command_test.dart:37]: Nit throughout the PR: "filename" is generally a single word in programming, so the variable would just be `filename`.

File: script/tool/test/update_excerpts_command_test.dart (hunk 1 above)
Comment (stuartmorgan) [reply script/tool/test/update_excerpts_command_test.dart:509]: There's not a conflict, the linter just gives a misleading suggestion for what you probably wrote as your other version. The version of this that's compliant with all of our lints is `testCases.forEach(runTest);`.

**Review by stuartmorgan — CHANGES_REQUESTED** [review]

File: script/tool/lib/src/update_excerpts_command.dart (hunk 2 above)
Comment (mike-v2) [reply script/tool/lib/src/update_excerpts_command.dart]: Shoot, there was a problem with my test case for `.cc` files. My mistake.

**Review by stuartmorgan — APPROVED** [review]
LGTM!

File: script/tool/lib/src/update_excerpts_command.dart [hunk 4]
```diff
@@ -146,6 +146,12 @@ class UpdateExcerptsCommand extends PackageLoopingCommand {
                   case '.cpp':
                     language = 'c++';
                     break;
+                  case '.m':
+                    language = 'objectivec';
+                    break;
```
Comment (tarrinneal) [inline script/tool/lib/src/update_excerpts_command.dart:151]: @stuartmorgan Seems like we don't have support for header files for either language. I'm not sure there's a simple solution though.

File: script/tool/test/update_excerpts_command_test.dart [hunk 5]
```diff
@@ -33,12 +33,16 @@ void runAllTests(MockPlatform platform) {
...
   });
 
-  Future<void> testInjection(String before, String source, String after,
-      {bool failOnChange = false}) async {
+  Future<void> testInjection(
+      {required String before,
+      required String source,
+      required String after,
+      required String filename,
+      bool failOnChange = false}) async {
```
Comment (tarrinneal) [inline script/tool/test/update_excerpts_command_test.dart:41]: nit: add a trailing comma `false,})`

**Review by tarrinneal — CHANGES_REQUESTED** [review]
Do you think there's a good way to handle .h files for c++ and objc?

**Review by tarrinneal — APPROVED** [review]
Changed my mind, merging now

File: script/tool/lib/src/update_excerpts_command.dart (hunk 4 above)
Comment (stuartmorgan) [reply script/tool/lib/src/update_excerpts_command.dart:151]: Good point. That would need either more special casing [here](https://github.com/flutter/packages/blob/dde5ebf5a4bad528882837eeb9fab1cf978947d8/script/tool/lib/src/update_excerpts_command.dart#L187-L191), or a more generalized solution (like a mapping from extensions to the set of allowed languages)).
